### DISCUSSION TITLE
Extending setuptools with finalize_distribution_options: How can I find the source directory

### DISCUSSION THREAD

**Discussion by jessie-ross** [pypa/setuptools#3959] [Q&A]
Hi there!

I'm trying to extend setuptools so it automatically brings in requirements from Pipfile.
I have it working but it's a bit hacky as I don't have a reliable reference to the source directory - I use the CWD, but that only works early on in the build with sdist, which is fine because the wheel gets built from the sdist.

Problems occur however if I try to install my package (pacakge A) locally from another package (package B), as the CWD will be set to package B.

The same issue occurs for [setuptools_scm](https://github.com/pypa/setuptools_scm), and I believe it is causing other issues there.

In the `finalize_distribution_options` entry point all I have access to is the `dist` object, is there a way I'm missing to get the source directory?

Thanks!

**Comment by abravalheri** [comment]
I usually go with `location = dist.src_root or os.curdir`.

> Problems occur however if I try to install my package (pacakge A) locally from another package (package B), as the CWD will be set to package B.

If setuptools is being called to build `package A`, the build frontend must call setuptools with CWD set to A as per PEP 517, even if the build was initiated from `package B`. I am not sure if I understood this sentence...

**Reply by jessie-ross** [reply]
Thanks for your reply, I'll see if I can create some code that replicates the issue.

> the build frontend must call setuptools with CWD set to A as per PEP 517

But I can say isn't always true, for example when building the wheel we are building from a temporary directory, and that temp dir is set to the CWD. This is specifically in `finalize_distribution_options`.

**Reply by abravalheri** [reply]
In the case of the building stage, I don't think setuptools itself has this information.

If the `sdist` is copied to a temporary directory, extracted and then the frontend runs `setuptools.build_meta.build_wheel`, effectively the source directory that setuptools uses is the one inside the temporary directory. Setuptools does not know where the source came from, it just know the source it is building at the moment.
